Fix bug with rendered empty elements in array

nachogiljaldo opened the pull request:
Related to https://github.com/Faleij/json-stream-stringify/issues/43

IIUC, there can be concurrent calls to _read which is why the ReadState is there (to try to avoid that case). However, because _read() is asynchronous, the nested call should do a wait, else, there is a risk of concurrent calls to _read that attempt to read.

Also, while I tried to understand the code, I broke a few things but the existing tests were not catching it, so I added a few new test cases.

nachogiljaldo committed:
Add tests and skip calls to _read() when still reading.

nachogiljaldo committed:
Await for _read.

Faleij commented:
I have reviewed your changes and the tests you have added passes without the code changes.
There is no need for an await on this._read; the readState ensures there can not be any simultaneous executions of _read.
I suspect there is something else causing the issue you have reported but it might be impossible to find without code that reproduces the error.

nachogiljaldo commented:
Hey @Faleij thanks for the prompt response!

> I have reviewed your changes and the tests you have added passes without the code changes.

Yes, sorry, I think I wasn't clear. What I meant is that while trying to explore a solution, I removed some code (basically everything related to `readState`) and all tests passed.
Then trying to reproduce my bug I added those test cases and they failed, so actually the original test suite was incomplete and this enhances it (it doesn't reproduce my problem, unfortunately). If you prefer another PR for that I'm happy to move them.

> There is no need for an await on this._read; the readState ensures there can not be any simultaneous executions of _read.
I suspect there is something else causing the issue you have reported but it might be impossible to find without code that reproduces the error.

IIUC, there's at least a few theoretical patches that could lead to this problem, _maybe_ they don't happen due to the nature of how nodejs works, but I don't think this await hurts.

Example:
- `_read` sets `readState` to `NotReading` and performs a "fire and forget" call to `_read`. First thing that call will do is setting the state to `Reading` to prevent concurrent calls. Let's say, like in my example, it now yields the CPU because it's reading from a `Readable` (in my case objects).
- the original call to `_read` then picks the CPU and sets readState to `NotReading` in spite of a call being waiting. This means we can end up with 2 concurrent calls if there's a `push` on the other (which will eventually be the case?).

Maybe that is unlikely to happen in the running model of nodejs, but I could see it potentially happen given there's ways _read can block and yield the CPU.

To be clear, I am not 100% convinced this is the problem I see (mainly because I could not reproduce), but it certainly is some sort of race-condition because I believe it depends on the amount of time spent blocking on reading on the `Readable`.

I am now out of office, currently I'm running a patched version on the library with some logs to try and find the problem, next week I will add more logs and test this solution.

nachogiljaldo committed:
Merge branch 'master' into fix_bug_with_rendered_empty_elements_in_array

nachogiljaldo committed:
Undo accidentally removing readState.

nachogiljaldo commented:
I will close as it was fixed.
